### ISSUE TITLE
Encountered pending RTE: 1321, a wave write;Index out of range for wave "ScaledData"

### ISSUE THREAD

**Issue by ru57y34nn** [AllenInstitute/MIES#1234]
While testing MIES release 2.4 we have encountered an error during data acquisition. So far this has only happened when running our X7Ramp stimulus, but MIES will suddenly stop acquisition and then we are unable to get the test pulse back or run any other stimuli. Igor is not frozen at this point and we are still able to save the experiment, but we are unable to acquire any more data once this issue occurs.

I will upload both affected experiment files from today as well as a copy of our stimulus set nwb file to the FTP server.
Filenames:
Gad2-IRES-Cre;Ai14-606845.10.01.01.pxp
Gad2-IRES-Cre;Ai14-606845.10.01.02.pxp
PS_Stimsets-20210824.nwb

Traceback information: 
  !!! Threadsafe assertion FAILED !!!
  Message: "Encountered pending RTE: 1321, a wave write;Index out of range for wave "ScaledData"."
  Please provide the following information if you contact the MIES developers:
  ################################
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  Stacktrace:
DQM_FIFOMonitor(...)#L94 [MIES_DataAcquisition_Multi.ipf]
AS_HandlePossibleTransition(...)#L103 [MIES_AcquisitionStateHandling.ipf]
AFM_CallAnalysisFunctions(...)#L113 [MIES_AnalysisFunctionManagement.ipf]
AssertOnAndClearRTError(...)#L4335 [MIES_Utilities.ipf]
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  Time: 2021-12-17T14:40:41-08:00
  Experiment: 2021_12_17_141417 (Packed)
  Igor Pro version: 9.0.0.10 (37840)
  ################################
  BUG AS_HandlePossibleTransition(...)#L80: The state transition AS_MID_SWEEP -> AS_POST_DAQ is not expected.

  !!! Threadsafe assertion FAILED !!!
  Message: "Encountered pending RTE: 963, Redimension;Can't change a locked wave."
  Please provide the following information if you contact the MIES developers:
  ################################
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  Stacktrace:
TPM_BkrdTPFuncMD(...)#L283 [MIES_TestPulse_Multi.ipf]
TS_GetNewestFromThreadQueue(...)#L33 [MIES_ThreadsafeUtilities.ipf]
AssertOnAndClearRTError(...)#L4335 [MIES_Utilities.ipf]

**Comment by t-b** [comment]
Thanks for the report. I'll have a look. Is the error reproducible?

**Comment by t-b** [comment]
I'm onto something and should have a fix ready by tomorrow (tuesday).

**Comment by t-b** [comment]
Me and @timjarsky looked at Gad2-IRES-Cre;Ai14-606845.10.01.01.pxp today again.
Here are our findings:
- The very last acquired sweep, was a ramp, but that did not get saved as sweep and thus does not show up in the DB. The data is completely acquired in root:MIES:HardwareDevices:ITC18USB:Device0:ScaledData.
- Only the analysis function RAMP entries are present as the stock ones are written after saving scaleData as sweep.
- The baseline QC failed but enough spikes were detected

#1236 only has mitigation against the consequences of the scaledData wave RTE above but no fix for it.

The bug could lie in the offset handling after restarting DAQ.

**Comment by ru57y34nn** [comment]
@t-b I have uploaded another experiment file to your server (Pvalb-IRES-Cre;Ai14-611951.05.01.01) which contains the same or a similar error. The rig operator was running the first sweep of the X7Ramp sweeps and acquisition automatically stopped and a Function Execution Error dialogue box appeared and I had the rig operator grab a screenshot (see attached image for details), after which we closed the dialogue box and immediately saved the experiment. 

The test pulse was stopped at this point but this time we were able to get the test pulse started again just by clicking the "Start Test Pulse" button, which I don't  believe worked last time. After the test pulse restarted we tried running the X7Ramp again but encountered the same error and we noticed that the Dashboard showed "Sweep not yet finished" as the result for X7Ramp. At this point we also noticed that the X6SP_Rheo_DA_0 sweeps had not passed so I had the rig operator rerun the X6SP_Rheo_DA_0 sweeps and we were able to get them to pass this time; however, I later noticed that in the attached screen shot, the second attempt at the X6SP_Rheo_DA_0 showed a "Pass" Result (circled in red) in the dashboard, but after attempting to rerun the X7Ramp the result for the second X6SP_Rheo_DA_0 had changed to "Sweep not yet finished". Let me know if you have any additional questions.
![Pvalb-IRES-Cre;Ai14-611951 05 01 01_error](https://user-images.githubusercontent.com/26012971/150033112-c401307a-c26e-4327-ace4-e9158a0216b2.jpg)

**Comment by t-b** [comment]
@ru57y34nn Thanks Rusty. Looks like the additional assertion triggered. Nice! I'll investigate and reply ASAP.

**Comment by t-b** [comment]
I've pushed a fix for the assertion to #1236. I would think that the other errors are consequences of the assertion but nevertheless will have a look tomorrow.

**Comment by t-b** [comment]
@ru57y34nn So I've updated #1236 to also include a better algorithm for determining if the sweep is really finished.

And it now looks like

![image](https://user-images.githubusercontent.com/1185677/150430620-3acd4f5a-b329-477d-9cb9-de1c00b54b61.png)

Regarding the unexpected passing I've created https://github.com/AllenInstitute/MIES/issues/1258 and tackle that ASAP.

**Comment by t-b** [comment]
I've added a patch for the case you found that invalid data was saved. SWS_SaveAcquiredData now checks more thoroughly, but also fails gracefully with BUG instead of ASSERT. In general it should be *very* rare to reach a state where SWS_SaveAcquiredData complains about invalid data.

I also made browsing invalid sweep/config wave combinations a non-fatal error. We can't display that data, but at least don't assert out.
